keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1197 @@
+identical cards:
+
+```
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+nodeName: worker-1
+driverName: pool.dra.example.com
+resources:
+- name: PoolOfLargeCards
+  attributes:
+  - name: meta
+    metadata:
+      labels:
+      - name: Configuration
+        value: Large
+  - name: Number
+    counter:
+      count: 8
+```
+
+If a driver needs to reduce resource capacity, then there is a risk that a claim gets
+allocated using that capacity while the kubelet is updating the
+NodeResourceSlices. The implementations of semantic models must handle
+scenarios where more resources are allocated than available. The DRA driver
+kubelet plugin must double-check that the allocated resources are still
+available when NodePrepareResource is called. If not, the pod cannot
+start until the resource comes back. Treating this as a fatal error during
+pod admission would allow to delete the pod and trying again with a new
+one.
+
+### Using semantic parameters as claim parameters
+
+The author of the `cards.dra.example.com` DRA driver in the example above might
+define a CRD like this:
+
+```
+kind: CardParameters
+apiVersion: dra.example.com/v1alpha1
+metadata:
+  name: my-parameters
+  namespace: user-namespace
+  uid: foobar-uid
+...
+spec:
+  minimumAPI: "v1.0.0"
+  supportsPowerSaving: true
+  requirements:
+    RAM: 10Gi
+    ComputeUnits: 2
+```
+
+Note that all fields can be fully validated, including value ranges that are
+specific to the underlying hardware.
+
+When a user creates an object of that type, the DRA driver control plane must
+convert it into an in-tree ResourceClaimParameters object because the scheduler
+doesn't understand `CardParameters`. That object then looks like this:
+
+```
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+
+metadata:
+  # This cannot be the same as my-parameters because parameter objects with a different
+  # type might also use it. Instead, the original object gets linked to below.
+  name: someArbitraryName
+  namespace: user-namespace
+
+generatedFrom:
+  name: my-parameters
+  kind: CardParameters
+  apiGroup: dra.example.com
+  uid: foobar-uid
+  generation: 10 # Ensures that only up-to-date generated parameters are used.
+
+parameters:
+- name: meta
+  metadata:
+    # A CELL expression with access to the labels, versions, and values defined for
+    # the node resource.
+    selector: versions["API"] >= "v1.0.0" && labels.has("SupportsPowerSaving")
+- name: RAM
+  counter:
+    required: 10000000
+- name: ComputeUnits
+  counter:
+    required: 2
+```
+
+The semantic is that all parameters must be satisfied by a single resource of
+the driver. Note that for the "pool of cards" example that makes it impossible
+to ask for two cards with no additional selection criteria and then get one
+from `PoolOfLargeCards` and one from a hypothetical second pool
+`PoolOfSmallCards` (not included in the example above).
```

Comment:
Simplifying the semantic like this is relevant for evaluating requirements during scheduling. Suppose we have two counters and a selector. With the current semantic, we can check each parameter against a resource and move to the next as soon as one resource does not match. 

If we allow "partial matches" (take x items from resource A and y items from resource B) then we need to define whether it's valid to take some items for counter M from resource A and some items for counter N from resource B. There may have to be some back-and-forth to satisfy all constraints. I'd like to this because the use cases are not that strong.

With the current semantic, "shared discrete cards" works and "single pool of cards" work. The latter is what people currently do with device plugin and (for selection) node labels.

We don't get "multiple pools with allocation from more than one pool". If we limit the evaluation to "one counter" + "metadata", then we could add a special case for it, but I'd prefer to punt on it for now.

Reply:
All of this goes away if we move the logic of how to define "pools" back down into the semantic model. @klueska is working on a proposal for such a model right now.